theano/tensor/opt.py

```diff
@@ -3502,6 +3507,18 @@ def local_adv_sub1_adv_inc_sub1(node):
             # investigate Alloc of 0s but with non constant shape.
             T.extract_constant(x, elemwise=False) != 0):
         return
+
+    if not inp.owner.op.set_instead_of_inc:
```

Comment:
This is too strict, but as it fix a bug I'll merge now. If we can know the indices aren't duplicate (like when they are constant), we could still do the optimization.

Not sure how much useful this is.